### PR TITLE
Chapter 2: Opening, Creeps sections

### PR DESCRIPTION
**Pull request by Shushishtok**
Godz needs some actual lines. And love. But mostly lines.

### CODE REVIEW

**Comment by RobinKa** [comment]
Currently broken because the golems arent alive to speak the dialog. Could require them and keep them in their default location (the center). Later we can come up with a better solution for the ghosts (or just change the default location to somewhere hidden).

**Comment by Shushishtok** [comment]
@RobinKa Fixed Golems missing from being required. They're now placed near the top Barracks.
The creep section is something I've spent more than a couple hours on trying to "stabilize" it when you constantly skip to it. In practice, it should never be possible in a game, as you should only be able to skip between Chapters.

It tries to properly clean and reinitialize the creep arrays, but sometimes it may still break on very specific cases.

File: common/general.d.ts [hunk 1]
```diff
@@ -15,20 +16,32 @@ declare const enum SectionName {
...
     Riki = "npc_dota_hero_riki",
-    GodzMudGolem = "npc_mud_golem_godz",
     TargetDummy = "npc_dota_tutorial_target_dummy",
 }
 
 declare const enum CustomAbilityKeys {
     CustomMiranaArrow = "custom_mirana_arrow",
 }
 
+declare const enum chapter2SpecificKeys {
```
Comment (Perryvw) [inline common/general.d.ts:33]: Do these types need to be in general or even common if they are chapter 2 specific? If they are not sent to panorama they should not be in common. Also this name is breaking our naming convention, should start with capital C.

File: game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+    ExecuteOrderFromTable({
+        OrderType: UnitOrder.ATTACK_MOVE,
+        Position: fightPosition,
+        UnitIndex: unit.entindex()
+    })
+}
+
+function createLaneCreeps(creepNames: string[], location: Vector, team: DotaTeam, sendCreepToFight: boolean, arrayToPush: CDOTA_BaseNPC[] | undefined): CDOTA_BaseNPC[] {
+    if (!arrayToPush) arrayToPush = [];
```
Comment (Perryvw) [inline game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts:327]: just make this a `const arrayToPush = []` and remove it from the parameters, it's not needed and just confusing.

File: game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+            }
+        }
+    }
+
+    const sniper = context[chapter2SpecificKeys.sniperEnemyHero]
+    if (sniper) {
+        if (IsValidEntity(sniper)) {
+            UTIL_Remove(sniper);
+        }
+        context[chapter2SpecificKeys.sniperEnemyHero] = undefined;
```
Comment (Perryvw) [inline game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts:287]: Suggestion for utils:
```ts
function removeContextEntityIfExists(context, entityKey) {
    const entity = context[entityKey];
    if (entity && IsValidEntity(entity)) {
        UTIL_Remove(entity);
    }
    context[entityKey] = undefined;
}
```

**Comment by RobinKa** [comment]
> They're now placed near the top Barracks.
I think they should be placed far away since they are now "ghosts"  and were killed off earlier.

**Comment by Shushishtok** [comment]
> I think they should be placed far away since they are now "ghosts" and were killed off earlier. Or make them invisible somehow like was mentioned on the discord (can do that later).

We can do it later, it doesn't really matter that much.

File: common/general.d.ts (hunk 1 above)
Comment (Shushishtok) [reply common/general.d.ts]: Moved to the chapter directory without `declare const` and imported on the chapter 2 portions of it.

File: game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts (hunk 2 above)
Comment (Shushishtok) [reply game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts]: Done, good idea.

File: game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts (hunk 3 above)
Comment (Shushishtok) [reply game/scripts/vscripts/Sections/Chapter2/SectionCreeps.ts]: Added this utility plus a check for arrays to take care of all instances in an array. Short testing seems to work as intended.

**Pull request merged**